### DISCUSSION TITLE
Manual Entry of TOTP Token Key

### DISCUSSION THREAD

**Discussion by kevingarman** [keycloak/keycloak#35527] [Q&A] [answered]
Hi,

I know that programmable hardware totp tokens (such as [token2 c301](https://www.token2.net/shop/product/token2-c301-i-programmable-hardware-token-iphone-compatible)) can be used with keycloak.  What I'm trying to determine is whether/how I could use a non-programmable hardware token (such as [token2 c202](https://www.token2.net/shop/product/token2-c202-hardware-token)).

I might be missing something, but it seems the only difference how the token's key is handled.  In the case of the programmable token, keycloak provides a key to be written to the token.  And, in the case of the non-programmable token, the key is provided by the manufacturer and would need to be input into keycloak.

Is there a way to specify to keycloak the key of the hardware token?

Thanks!

**Comment by kevingarman** [comment]
....my intent would be the following scenario.

--I have a box of hardware tokens for which I know all the keys (ie, token1 key is 123, token2 key is 456, etc)
--I send the box of tokens to a the end user and they can issue them (ie, joe received token1 and sam received token2, etc)
--I can manually bind the appropriate token to the each user through the keycloak admin interface
--users don't have to setup anything

**Comment by eminhuseynov** [comment, marked as answer]
Have a look here as well https://github.com/keycloak/keycloak/pull/21117

**Reply by kevingarman** [reply]
Thanks, I finally got back to testing this.  Editing the database per issue 21117 does work.  So then....

--can keycloak be set to periodically refresh itself from the database (so I don't have restart keycloak upon db edit)?

--can I enter the base32 value from the keycloak web interface somehow?
